[1~3단계 - 방탈출 사용자 예약] 강산(조재중) 미션 제출합니다.

m-a-king opened the pull request:
안녕하세요, 찰리!
이번 미션은 나이스한 페어, 시소와 함께 제 코드로 즐겼습니다.

찰리의 모든 피드백을 기대하고 있겠습니다! 정말 감사합니다.

# [> 이번 미션 변경 사항 바로 가기 <](https://github.com/woowacourse/spring-roomescape-member/pull/222/files/77d35d6d545a9f5a9ecd27c7d473bf342936f5ff..a4abd3d287f7bd6d53abd72bbbec7146c57da17f)

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

https://prolog.techcourse.co.kr/studylogs/4271

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [x] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

- [x] 1. 한 메서드에 오직 한 단계의 들여쓰기만 한다

- [ ] 2. else 표현을 사용하지 않는다 -> 테스트에서 사용함

- [ ] 3. 모든 원시 값과 문자열을 포장한다 -> int count를 포장하지 않음

- [ ] 4. 한 줄에 점을 하나만 사용한다 -> stream에서 둘 이상 사용함

- [ ] 5. 이름을 줄여 쓰지 않는다(축약 금지) -> ByParams라고 적음

- [ ] 6. 모든 엔티티를 작게 유지한다 -> 3개 이상의 필드를 가지는 엔티티

- [x] 7. 3개 이상의 기본 데이터타입(Int, String, Double 등) 프로퍼티를 가진 타입을 구현하지 않는다

- [ ] 8. 일급 콜렉션을 사용한다

- [ ] 9. getter/setter를 구현하지 않는다

2/9 = 0.22222... -> 0.2 -> 1점 😿 

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [x] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

## DB 엔티티와 도메인 엔티티를 구분하는 것이 좋은 방향인지에 대해 고민

처음에는 DB 테이블 구조에 맞춰 DB 엔티티를 만들고, 도메인 엔티티는 따로 분리하여 리포지토리 인터페이스를 통해 데이터를 추상적으로 가져오는 구조로 설계했습니다. 하지만 실제 구현에서는 예약 시간이나 테마 같은 객체들을 id만 들고 있게 하고 필요할 때 추가로 쿼리를 날리기보다는, 처음부터 조인으로 함께 조회하는 것이 성능상 유리했고, 그렇게 되면서 DB 엔티티와 도메인 엔티티의 구조가 거의 유사해졌습니다.

현재는 의존성 방향을 분리하기 위해 두 구조를 유지하고 있지만, 구조적으로 큰 차이가 없다 보니 굳이 이렇게 나누는 것이 의미 있는지에 대한 고민이 생겼습니다. 이런 상황에서도 도메인과 DB 모델을 분리하는 것이 여전히 유효한 설계라고 보시는지, 혹은 제가 현 상황을 잘못 이해하고 있는 부분이 있는지 조언을 부탁드리고 싶습니다.

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

---
---
---

# (2025-05-05 21:21) 추가 질문이 없습니다!

# [리뷰 이후 변경 사항](https://github.com/woowacourse/spring-roomescape-member/pull/222/files/a4abd3d287f7bd6d53abd72bbbec7146c57da17f..af3caccdfb3b3139724c57a3f47760033cab79c0)

# [리뷰 이전 변경 사항 + 리뷰 이후 변경 사항](https://github.com/woowacourse/spring-roomescape-member/pull/222/files/77d35d6d545a9f5a9ecd27c7d473bf342936f5ff..af3caccdfb3b3139724c57a3f47760033cab79c0)

solarsalt committed:
docs: add PR template

m-a-king committed:
[방탈출 예약 관리]

m-a-king committed:
docs: 예약 생성/삭제 검증 명세 추가

m-a-king committed:
refactor: 예약 정적 팩토리 메서드 개선

m-a-king committed:
feat: 참조 중인 예약 시간 삭제 불가 기능 추가

m-a-king committed:
feat: 과거 시간/날짜 예약 생성 검증 기능 추가

m-a-king committed:
feat: 사용자 입력 예외 처리 (이름, 날짜, 시간)

m-a-king committed:
refactor: time 조회 책임 이동

m-a-king committed:
feat: 중복 시간 추가 불가 검증 추가

m-a-king committed:
feat: 예약자명 모든 유니코드 공백 검증 추가

m-a-king committed:
feat: 중복 예약 검증 기능 추가

m-a-king committed:
feat: 전역 예외 핸들링 추가 (IAE, ISE)

m-a-king committed:
docs: 테마 추가

m-a-king committed:
feat: 테마 페이지 컨트롤러 추가

m-a-king committed:
feat: 테마 도메인 엔티티 추가, 관련 VO 추가

m-a-king committed:
feat: 테마 전체 조회 기능 추가

m-a-king committed:
refactor: 코드 일관성 개선

m-a-king committed:
feat: 테마 생성, 삭제 기능 추가

m-a-king committed:
refactor: 패키지 이동

m-a-king committed:
feat: 테마 동등성 설정

m-a-king committed:
test: 테마 저장/삭제 테스트 추가

m-a-king committed:
test: 테마 조회 테스트 추가

m-a-king committed:
rafactor: 존재 여부 확인 메서드 명 개선

m-a-king committed:
feat: 예약 관련 엔티티, DTO에 테마 추가

m-a-king committed:
refactor: 예약 시간 객체 생성 방식 개선

m-a-king committed:
feat: 예약 객체에 테마 추가

m-a-king committed:
refactor: 패키지 이동

m-a-king committed:
feat: 쿼리 컬럼 명시

m-a-king committed:
refactor: 타임리프 프리픽스 수정

m-a-king committed:
docs: 리드미에 예약 관련 기능 추가

m-a-king committed:
feat: 응답에 따른 화면 렌더링 기능 추가

m-a-king committed:
feat: 날짜와 테마에 따른 예약 가능 여부를 포함한 시간 조회 기능 추가

m-a-king committed:
refactor: 값 객체 사용

m-a-king committed:
feat: 사용자 페이지 컨트롤러 추가

m-a-king committed:
feat: 테마 랭킹 조회 기능 추가

m-a-king committed:
feat: 예약 생성 예외 메시지 추가

m-a-king committed:
refactor: 계층 간 전달하는 인자 컨벤션 개선

m-a-king committed:
refactor: 코드 일관성 개선

m-a-king committed:
refactor: 명명 개선

Ryan-Dia commented:
@coderabbitai full review

Gomding commented on README.md:
> 처음에는 DB 테이블 구조에 맞춰 DB 엔티티를 만들고, 도메인 엔티티는 따로 분리하여 리포지토리 인터페이스를 통해 데이터를 추상적으로 가져오는 구조로 설계했습니다. 하지만 실제 구현에서는 예약 시간이나 테마 같은 객체들을 id만 들고 있게 하고 필요할 때 추가로 쿼리를 날리기보다는, 처음부터 조인으로 함께 조회하는 것이 성능상 유리했고, 그렇게 되면서 DB 엔티티와 도메인 엔티티의 구조가 거의 유사해졌습니다.
> 
> 현재는 의존성 방향을 분리하기 위해 두 구조를 유지하고 있지만, 구조적으로 큰 차이가 없다 보니 굳이 이렇게 나누는 것이 의미 있는지에 대한 고민이 생겼습니다. 이런 상황에서도 도메인과 DB 모델을 분리하는 것이 여전히 유효한 설계라고 보시는지, 혹은 제가 현 상황을 잘못 이해하고 있는 부분이 있는지 조언을 부탁드리고 싶습니다.

현재 구조를 유지하는것에 강산은 어떤 장점이 있다고 생각하는지 궁금해요!
현재구조에서 필드가 추가되거나 수정될 때 유지보수에 불편함은 없을까요?

m-a-king replied:
리포지토리는 응답 DTO가 아닌 도메인 객체를 반환하고, 응답 DTO를 만들기 위해서는 이 도메인 객체가 null-safe하게 설계되어 있기 때문에 내부 필드가 비어 있어선 안 됩니다.

즉, 응답에 당장 사용되지 않는 필드라 하더라도, 도메인 객체를 생성하려면 모든 값을 반드시 채워야 하고, 그 결과 자연스럽게 조인을 하지 않을 수 없는 상황이 만들어집니다. 필요한 값들을 개별적으로 n번 조회하기보다는, 처음부터 조인을 활용해 한 번에 데이터를 불러오는 방식이 훨씬 효율적이기 때문입니다.

결국 도메인 객체가 완전성을 요구하는 한, 레포지토리 내부에서는 어차피 조인을 하게 되고, 이런 구조에서는 엔티티 계층을 나누더라도 조인이 필연적으로 요구되면서 DB 테이블과 도메인 모델이 다시 밀접해지는 아이러니가 발생합니다.

사실 저는 처음에 DB 엔티티와 도메인 엔티티를 명확히 분리하고, 도메인 중심의 설계를 통해 더 유연하고 확장 가능한 구조를 만들고 싶었습니다. 하지만 쿼리 최적화와 객체 생성 제약을 고려하다 보니, 결국 도메인 객체 역시 테이블 구조와 닮아가게 되었고, 처음에 의도했던 분리의 효과는 점점 흐려졌습니다.

자연스럽게 다음과 같은 질문들이 생겼습니다.
- 도메인 엔티티과 DB 엔티티를 구분하려는 시도 자체가 잘못된 방향은 아닐까?

- 비용보다 가독성과 명시적 구조를 더 중요하게 본다면, 조인을 사용하지 않고 각 DB 엔티티를 나눠 가져와 도메인 객체를 조립하는 방식도 가능할까?

- 도메인 객체의 null-safe 검증이 부담이라면, 애초에 리포지토리에서 자신만의 응답 DTO를 반환하는 설계가 더 적절하지 않을까? 그렇게 된다면 불필요한 정보를 도메인 객체로 매핑하지 않고, 정말 필요한 값만 쿼리해서 응답에 맞게 가공할 수 있을 테니까. 다만 도메인 객체의 입장에서 만들어진 리포지토리가 필요한 값만을 출력하는 것은 구조적으로 어긋난다는 생각도 듭니다. 리포지토리를 바라보는 시선 자체를 변경한다면, 가능할 것이지만요...

결국 이 질문은 더 근본적인 고민으로 이어집니다.

설계 원칙은 분명 유연성과 유지보수성을 가져다줍니다. 하지만 그 가치는 결국 비용을 감수할 여지가 있을 때만 실현됩니다.
도메인 엔티티와 DB 엔티티를 구분하는 행위 자체가 성능, 복잡도, 개발 리소스 측면에서 감당 가능한 수준을 넘어서게 되면, 그 시점부터는 오히려 과잉 구조로 인식됩니다.

그래서 지금의 제 구조 역시, 초기에는 의미 있는 분리라고 생각했지만, 지금 와서는 오히려 불필요한 설계로 보일 수도 있다는 생각이 듭니다. 특히 단순한 CRUD 위주의 어플리케이션에서는 도메인 모델이 자연스럽게 데이터베이스 구조를 따라가게 되고, 그 흐름이 오히려 더 현실적이고 합리적인 선택처럼 느껴졌습니다.

m-a-king replied:
> 현재 구조를 유지하는것에 강산은 어떤 장점이 있다고 생각하는지 궁금해요!
현재구조에서 필드가 추가되거나 수정될 때 유지보수에 불편함은 없을까요?

1. 현재 구조를 유지하는 장점은

- 구조적으로 도메인 엔티티와 DB 엔티티을 구분해두면, JPA 어노테이션이 가득한 객체나 DB 테이블에 맞춘 필드명을 도메인 로직에 그대로 끌고 들어오지 않아도 된다는 장점이 있다고 생각합니다.
- 만약 도메인 객체에 실제 로직이 들어가기 시작한다면, 이를 그대로 DB 엔티티로 쓰는 건 부담스러울 수 있기 때문에, 사전에 구조를 나눠두는 건 이후 확장을 위한 안전장치로 작용할 수 있다고 생각합니다.

---

2. 유지보수 관점에서는

- 구조가 분리되어 있는 만큼, 필드 추가·수정 시 DB 엔티티, 도메인 객체, 매핑 코드, 테스트 코드 모두 수정해야 하는 부담이 분명히 존재합니다.
- 하지만 그만큼 변경이 미치는 범위가 명확히 드러나고, 각 계층의 책임을 인지한 채로 수정을 진행할 수 있다는 점에서 초기 설계만 잘 되어 있다면 오히려 추적과 관리에는 도움이 될 수도 있다고 생각합니다

Gomding replied:
JPA 를 사용하지 않고 있는데 JPA Entity 객체 얘기를 끌어오는것은 과하다고 생각해요!
지금은 현재 구현된 Reservation 객체 상태를 두고 얘기해보면 좋겠어요 :)
(추가로 말씀드리면 JPA 엔티티 객체도 도메인 객체로 사용하는것을 유도하고 있습니다)

- https://docs.spring.io/spring-data/jpa/reference/repositories/core-concepts.html
- https://www.inflearn.com/community/questions/90087/domain%EA%B3%BC-repository-%EA%B5%AC%ED%98%84-%EC%83%81%EC%84%B8%EB%A5%BC-%EB%B6%84%EB%A6%AC%ED%95%98%EA%B3%A0%EC%9E%90-%ED%95%A0-%EB%95%8C-entity-%EB%94%94%EC%9E%90%EC%9D%B8

> 만약 도메인 객체에 실제 로직이 들어가기 시작한다면, 이를 그대로 DB 엔티티로 쓰는 건 부담스러울 수 있기 때문에, 사전에 구조를 나눠두는 건 이후 확장을 위한 안전장치로 작용할 수 있다고 생각합니다.

도메인 객체를 그대로 DB 엔티티로 사용하는게 부담스러운 이유가 궁금해요!

> 하지만 그만큼 변경이 미치는 범위가 명확히 드러나고, 각 계층의 책임을 인지한 채로 수정을 진행할 수 있다는 점에서 초기 설계만 잘 되어 있다면 오히려 추적과 관리에는 도움이 될 수도 있다고 생각합니다

수정 시 두 군데를 동시에 관리해야한다는 점은 생각보다 힘든 작업이 될 수 있어요
심지어 형태가 동일하다면 그 장점이 조금은 희석되는것도 있죠
강산의 말대로 분리했을때 장점도 명확하지만 단점도 명확하다고 생각합니다 :)
분리 했을 때 장점이 정말 명확한지 의문이 여전히 있어요!

만약 도메인 객체가 복잡하고 DB 와 매핑과정이 복잡하고 형태가 많이 다르다면 
강산의 설계의 장점이 훨씬 부각될 것 같은데
현재는 단순한 형태의 도메인 객체라 와닿지 않는 것일수도 있겠네요 :)
만약 복잡한 도메인 객체가 나타났을때만 분리하는 선택을 하는것은 어떻게 생각하시나요?
일관성있는 구조도 좋지만 저는 일부를 위해 전체가 복잡해지는것은 피하고싶다는 의견입니다 😃

m-a-king replied:
> JPA 를 사용하지 않고 있는데 JPA Entity 객체 얘기를 끌어오는것은 과하다고 생각해요!
지금은 현재 구현된 Reservation 객체 상태를 두고 얘기해보면 좋겠어요 :)
(추가로 말씀드리면 JPA 엔티티 객체도 도메인 객체로 사용하는것을 유도하고 있습니다)

맞습니다. 아직 JPA를 도입하지 않은 상황에서 JPA를 논의에 끌어오는 것은 논점이 흐려질 수 있습니다.
저도 JPA 엔티티를 도메인 객체로 사용하는 것이 현실적인 선택이라는 글들을 많이 접했었고, 참조해 주신 문서들도 큰 도움이 되었습니다. 김영한 님과 [최범균 님](https://www.inflearn.com/community/questions/946595/%EB%8F%84%EB%A9%94%EC%9D%B8%EA%B3%BC-jpa-%EC%97%94%ED%8B%B0%ED%8B%B0)의 커뮤니티 글은 읽어본 적이 있었지만, 공식 문서는 이번에 처음 확인하게 되었네요. 조금 부끄럽고, 반성하게 됩니다.

> 도메인 객체를 그대로 DB 엔티티로 사용하는게 부담스러운 이유가 궁금해요!

저도 제 스스로의 판단이 모순처럼 느껴졌고, 왜 그렇게 생각했는지 돌아보게 되었습니다. 아마도 예전부터 경험하고 싶다고 생각해온 헥사고날 아키텍처 스타일의 구조의 일부를 시도한 영향이 컸던 것 같아요. 외부 의존성과 무관한 순수한 도메인 객체를 만들고 싶다는 생각이 있었고, 이를 무리하게 적용했던 것 같습니다. 사실 이미 롬복을 적극 사용하고 있는 상황 자체부터 잘못된 것이 아닌가? 생각도 들었습니다.

> 수정 시 두 군데를 동시에 관리해야한다는 점은 생각보다 힘든 작업이 될 수 있어요
심지어 형태가 동일하다면 그 장점이 조금은 희석되는것도 있죠
강산의 말대로 분리했을때 장점도 명확하지만 단점도 명확하다고 생각합니다 :)
분리 했을 때 장점이 정말 명확한지 의문이 여전히 있어요!
만약 도메인 객체가 복잡하고 DB 와 매핑과정이 복잡하고 형태가 많이 다르다면
강산의 설계의 장점이 훨씬 부각될 것 같은데
현재는 단순한 형태의 도메인 객체라 와닿지 않는 것일수도 있겠네요 :)

맞습니다. 실제로 도메인 객체와 DB 모델 간 형태가 거의 유사한 상황에서 이 둘을 나누는 것은 분리라는 구조적 장점보다 오히려 중복 관리의 부담이 더 크게 느껴졌던 것 같습니다. 제가 원했던 설계는 나름의 의미가 있었지만, 지금 와서 보면 그 장점은 이미 많이 사라진 상태고, 현재 문제 크기에 비해 해결 방법이 과도했다는 생각이 듭니다.

> 만약 복잡한 도메인 객체가 나타났을때만 분리하는 선택을 하는것은 어떻게 생각하시나요?
일관성있는 구조도 좋지만 저는 일부를 위해 전체가 복잡해지는것은 피하고싶다는 의견입니다 😃

완전히 동의합니다. 지금까지의 고민을 통해 결국 저도 같은 결론에 다다른 것 같습니다. 처음부터 구조를 이상적으로 만들기보다는, 현재 요구사항에 맞는 현실적 구조를 선택하고, 필요해졌을 때 점진적으로 분리해 나가는 것이 더 합리적이라는 걸 느꼈습니다. 일관성도 중요하지만, 전체 구조가 일부를 위해 복잡해지는 것은 피하고 싶다는 의견에 공감합니다. 

---

그렇지만 지금 직접 경험해보는 것이 너무 중요한 것 같아요. 지금이 아니라면 언제 할 수 있을까 하는 생각이 자주 들었어요... 😄 
DB 엔티티와 도메인 엔티티를 구분하며 장단점을 직접 체감해서 너무 기분이 좋네요!!!

Gomding commented on src/main/java/roomescape/reservation/infrastructure/H2ReservationRepository.java:
```diff
@@ -0,0 +1,253 @@
+
+@Repository
+@RequiredArgsConstructor
+public class H2ReservationRepository implements ReservationRepository {
```
해당 클래스 내부에 H2에 대한 내용이 존재할까요?

m-a-king replied:
현재 H2 접두어를 가진 구현체들 내부에는 H2에만 의존적인 코드는 없습니다.
명시적으로 H2라는 이름을 붙인 이유는, 보편적으로 사용하는 MySQL과 달리 H2는 `INSERT ... ON DUPLICATE KEY UPDATE` 같은 Upsert 문법을 지원하지 않기 때문입니다. H2에서는 이를 `MERGE` 키워드로 처리한다는 점은, 이전 레벨 1 장기 미션에서 MySQL과 H2를 동시에 사용하며 학습했던 내용입니다.

다만 현재 이 클래스는 아직 Update 관련 기능을 포함하고 있지 않기 때문에, 지금 시점에서 H2라는 이름을 붙이는 것이 적절한가에 대한 고민이 생기고 있습니다.
나아가 향후 Upsert 기능이 도입되었을 때도, 특정 DBMS 이름보다는 구현 방식에 초점을 맞춰 JdbcTemplateReservationRepository처럼 명명하는 것이 더 자연스러울 수도 있겠다는 생각이 들었습니다.

Gomding replied:
말씀대로 DB 벤더에 따라 구현체를 만들 수 있지만
만약 벤더가 바뀌더라도 SQL 문자열만 달라지는 정도로 유지보수하면 끝날것 같아요!

하지만 jdbcTemplate 같은 DB 접근기술이 변경된다면 Repository 구현 전체가 변경되겠죠 😃 
해당 관점에서 본다면 말씀대로 `JdbcTemplateReservationRepository` `JdbcReservationRepository` 같은 이름이 더 어울릴것 같아요!

m-a-king replied:
맞습니다! 빠르게 수정하겠습니다.

Gomding commented on src/main/java/roomescape/reservation/application/ReservationService.java:
```diff
@@ -0,0 +1,19 @@
+import java.time.LocalDate;
+import java.util.List;
+
+public interface ReservationService {
```
`ReservationService` 를 추상화 하신 배경이 궁금해요! 
대부분 interface 를 두고 구현체를 만드셨는데 어떤 장점이 있어서 이런 구조로 만들어주셨나요?

강산의 추상화 기준이 궁금해요!

m-a-king replied:
다음과 같은 경우에 (인터페이스 기반의) 추상화를 적용하는 편입니다

1. 도메인 객체 내부의 전략이나 규칙이 자주 바뀔 가능성이 있을 때, 변화에 유연하게 대응하기 위해 구현을 추상화합니다.
2. 전체적인 아키텍처를 설계하며 계층 간의 의존 방향을 명확히 하고자 할 때, 상위 계층이 하위 구현에 의존하지 않도록 경계를 설정하는 수단으로 인터페이스를 도입합니다.

예를 들어, 저는 ReservationService 인터페이스를 단순히 구현체에서 추상화한 것이 아니라, 컨트롤러의 입장에서 기대하는 동작을 명세한 계약으로 바라보고 작성했습니다. 즉, 구현을 위한 추상화가 아니라 호출자가 바라보는 시점에서 정의된 추상화였고, 그에 따라 구현체가 작성된 구조입니다.

이러한 port-first 구조는 마치 TDD처럼 무엇을 할 것인가?를 먼저 정의(고정)하고 구현을 따라가게 하며, 각 계층이 자신의 기대 범위만 의존하게 함으로써 사고 범위를 명확히 제한하고, 현재 문제에 집중할 수 있도록 해주는 구조적 장치가 될 수 있었습니다.

이 방식을 통해 각 계층은 자신이 기대하는 동작만 의존하게 되며, 구현 세부에 대한 고민 없이도 아키텍처의 흐름을 그려갈 수 있습니다.

정리하자면, 제 추상화 기준은 다음 두 가지에 중점을 둡니다
- 복잡도를 숨기기 위한 추상화보다는, 기대 동작을 선명하게 드러내기 위한 추상화
- 교체 가능성이라는 불확실성보다, 현재 시점에서 계층 간 경계와 책임을 명확히 하기 위한 추상화

모든 객체에 무조건 인터페이스를 붙이는 것은 설계에 대한 불안감에서 비롯된 오버엔지니어링일 수 있습니다. 하지만 전체 구조를 먼저 설계해 나갈 때는 인터페이스가 각 계층의 시야를 제한하고 책임을 분리해주는 도구가 되었습니다.

Gomding replied:
자세하게 답변주셔서 감사해요!
인터페이스를 붙이는데 충분한 이유가 있다고 생각드네요 :)

추상화를 하는 것은 말씀대로 경계를 명확하게 하고 유연함을 줄 수 있지만
복잡도 증가라는 단점은 있다고 생각해요
그런 부분을 트레이드 오프하셨다고 생각하는데
설계를 지키기위해 항상 인터페이스를 두고 구현하는게 맞을지는 미션을 진행하면서 계속 고민해보셨으면 좋겠어요!
인터페이스를 생략해도 될 때를 적절하게 판단하는 기준을 정해보셔도 좋구요 :)

m-a-king replied:
감사합니다!

현재까지 레벨2에서 이 주제가 가장 재미있어서 저만의 기준으로 블로그에도 정리해볼 것 같습니다 😄

Gomding commented on src/main/java/roomescape/reservation/application/ReservationServiceImpl.java:
```diff
@@ -0,0 +1,58 @@
+@RequiredArgsConstructor
+public class ReservationServiceImpl implements ReservationService {
+
+    private final ReservationQueryUseCase reservationQueryUseCase;
+    private final ReservationCommandUseCase reservationCommandUseCase;
```
Query 와 Command 로 usecase 를 분리하신 이유가 궁금해요!
Repository 를 바로 의존하는것과 비교해서 어떤 장점이 있나요?

m-a-king replied:
먼저 네이밍에 대해 간단히 설명드리자면, 현재 구조에서 Service는 컨트롤러 단에서 호출되는 주요 서비스(ApplicationService)입니다. 
QueryUseCase와 CommandUseCase는 실질적으로는 QueryService, CommandService에 가까운 구조입니다. UseCase라는 이름은 단일 책임을 의미할 수 있어 다소 혼동을 줄 수 있고, 실제로는 여러 유즈케이스를 묶은 읽기 혹은 쓰기 파사드에 가깝습니다. 그래서 Reader, Executor, 혹은 Service라는 이름으로 변경하는 것도 고려하고 있습니다.

---

쿼리와 커맨드 단위로 분리한 이유는 크게 세 가지입니다.

1. 순환 참조를 방지하기 위해서입니다.
조회와 명령 기능이 하나의 서비스에 함께 있을 경우 구성상 서로를 참조하게 되기 쉬운데, 이를 구조적으로 분리함으로써 의존을 끊고 설계를 단순화할 수 있습니다.

2. 호출자 입장에서 필요한 기능만 의존할 수 있도록 하기 위해서입니다. 
조회만 필요한 경우는 Query 인터페이스만, 명령만 필요한 경우는 Command 인터페이스만 사용하면 됩니다. (ISP)

3. 유즈케이스를 더 세분화하기 쉬워지기 때문입니다. 지금은 Query나 Command 단위로 묶여 있지만, 추후에는 그 안에 더 작게 쪼개진 UseCase들이 들어가게 될 가능성을 고려했습니다.. 이 흐름을 염두에 두고 계층을 나눠두면 관리가 훨씬 수월해질 거라 생각했습니다.

---

Repository를 직접 의존하지 않고 UseCase 레이어를 두는 이유는, Repository는 데이터 접근을 위한 기술적 계층이고, UseCase는 비즈니스 흐름을 표현하는 계층이기 때문입니다. Repository를 직접 호출하면 로직이 여기저기 흩어질 수 있고 재사용도 어려워집니다. UseCase를 통해 비즈니스 흐름을 한 곳에 응집시키면, 의미 단위로 로직을 정리할 수 있어 설계가 훨씬 명확해진다고 생각했습니다.

정리하자면, Query와 Command로 나눈 이유는 구조적인 순환 의존을 막기 위한 현실적인 이유부터, 호출자 중심의 책임 분리와 이후 더 세분화된 유즈케이스 구성을 고려한 결과입니다.

Gomding replied:
> 순환 참조를 방지하기 위해서입니다.
조회와 명령 기능이 하나의 서비스에 함께 있을 경우 구성상 서로를 참조하게 되기 쉬운데, 이를 구조적으로 분리함으로써 의존을 끊고 설계를 단순화할 수 있습니다.

서비스 끼리는 순환 참조를 방지하기 위해
`한 쪽 방향으로 의존만 허용한다` 라는 규칙으로도 해결할 수 있을 것 같은데
강산은 이 부분을 어떻게 생각하는지 궁금해요!

> Repository를 직접 의존하지 않고 UseCase 레이어를 두는 이유는, Repository는 데이터 접근을 위한 기술적 계층이고, UseCase는 비즈니스 흐름을 표현하는 계층이기 때문입니다

Repository 를 데이터 접근 기술계층이라고 하시기엔 Repository 인터페이스는 domain 에 위치하신것 같아요!
그럼 Repository 는 도메인 계층으로써 도메인 객체의 컬렉션을 가지고 있는 것이고
Service 에서는 데이터 접근을 위한 기술적 계층이라는 것을 모르고있는것 아닐까요?

> Repository를 직접 호출하면 로직이 여기저기 흩어질 수 있고 재사용도 어려워집니다. UseCase를 통해 비즈니스 흐름을 한 곳에 응집시키면, 의미 단위로 로직을 정리할 수 있어 설계가 훨씬 명확해진다고 생각했습니다.

동의합니다 👍 👍 

> 유즈케이스를 더 세분화하기 쉬워지기 때문입니다. 지금은 Query나 Command 단위로 묶여 있지만, 추후에는 그 안에 더 작게 쪼개진 UseCase들이 들어가게 될 가능성을 고려했습니다.. 이 흐름을 염두에 두고 계층을 나눠두면 관리가 훨씬 수월해질 거라 생각했습니다.

현재는 해당 구조로인해 복잡도가 올라갔을 것 같은데
트레이드 오프라고 생각하셨을까요!

m-a-king replied:
> 서비스 끼리는 순환 참조를 방지하기 위해
한 쪽 방향으로 의존만 허용한다 라는 규칙으로도 해결할 수 있을 것 같은데
강산은 이 부분을 어떻게 생각하는지 궁금해요!

맞습니다. 서비스 간 순환 참조를 방지하기 위해 단방향 의존 규칙을 세우는 것도 하나의 방법이라고 생각합니다.
예를 들어 현재 상황에서는, Time을 삭제할 때 Reservation에 참조되고 있는지를 직접 확인하지 않고, 삭제 시 발생할 수 있는 외래 키 제약 예외를 포착해 처리하는 방식으로도 해결할 수 있을 것입니다.

또는 유즈케이스의 수준을 나눠서, 도메인 로직을 수행하는 유즈케이스와 단순 CRUD만 처리하는 유즈케이스로 구분한 뒤, 후자를 의존 대상으로 삼아 단방향 의존을 유지하는 구조도 가능합니다. 
물론 유즈케이스를 구분한다고 해서 도메인 간 참조 자체가 완전히 사라지는 것은 아닐 것 같습니다. 특히 CRUD 수준에서의 도메인간의 의존은 여전히 존재하게 되며, 도메인 로직이 서로 얽히는 상황도 발생할 수밖에 없을 것 같습니다. 이런 결합을 구조적으로 끊어내기 위해서는, 결국 이벤트 기반 아키텍처 같은 느슨한 연결 방식이 필요한 시점이 오지 않을까 싶습니다.

아직은 확신할 수 없지만, 저는 문제 상황에 완전히 직면한 후 해결책을 찾기보다는 예견 가능한 수준의 문제를 해결해낼 수 있는 구조를 만들어두고, 이후 문제가 발생했을 때 일관성이 조금 무너지더라도, 가장 적은 비용으로 해결할 수 있는 방법을 선택하는 것이 현실적인 방향이라고 생각합니다.

> Repository 를 데이터 접근 기술계층이라고 하시기엔 Repository 인터페이스는 domain 에 위치하신것 같아요!
그럼 Repository 는 도메인 계층으로써 도메인 객체의 컬렉션을 가지고 있는 것이고
Service 에서는 데이터 접근을 위한 기술적 계층이라는 것을 모르고있는것 아닐까요?

맞습니다. 어째서 제가 Repository를 데이터 접근 기술 계층이라고 설명했을까요... 죄송합니다. Repository는 일종의 컬렉션처럼 관리하고 있다고 정정하겠습니다. 

> 현재는 해당 구조로인해 복잡도가 올라갔을 것 같은데
트레이드 오프라고 생각하셨을까요!

네. 복잡도가 올라가는 것을 확실히 느꼈습니다. 현재 상황에서 문제 해결에만 집중했다면 서비스에서 각 도메인들의 Repository를 참조할 수 있었고, 그렇게 했다면 적은 복잡도로 문제를 해결할 수 있었습니다. 그치만 문제를 일반화하고 어떻게 해결하는 것이 더 좋은 방법일까 고민하며 트레이드 오프를 고려해서 구현했습니다!

Gomding commented on src/main/java/roomescape/reservation/application/converter/ReservationConverter.java:
```diff
@@ -0,0 +1,62 @@
+
+import java.util.List;
+
+public class ReservationConverter {
```
해당 Converter는 책임이 너무 많은것 같은데 어떻게 생각하시나요?
무엇을 변환하는지 예측이 가능하실까요 🤔 

각 클래스에서 변환을 담당해도 될 것 같은데 Converter 를 구현하신 이유가 궁금해요!

m-a-king replied:
맞습니다. 처음에는 Request/Response DTO와 DB 엔티티 각각이 도메인 엔티티로의 변환 책임을 가지고 있었습니다.
하지만 도메인 엔티티가 원시값이 아닌 VO 기반으로 구성되면서 생성 방식이 다양해졌고, 변환 경로도 복잡해졌기 때문에, 당시에는 변환을 한 곳에 모아 관리하는 컨버터 클래스를 두는 것이 더 명시적이고 구조적으로도 나은 방식이라고 판단했습니다.

이후 도메인 엔티티가 완성된 VO만을 받도록 구조를 정리하면서, 컨버터의 필요성을 다시 돌아보게 되었습니다.
이 과정에서 자연스럽게 DTO나 DB 엔티티가 변환 책임을 가져도 되는가에 대한 고민이 뒤따랐고, 특히 이들이 도메인을 의존하는 구조가 괜찮은지에 대한 확신이 서지 않아, 기존 컨버터 중심 구조를 유지했던 측면도 있었습니다.

컨버터 클래스에 변환 로직을 집중시킴으로써 얻는 이점도 있었지만, 말씀해주신대로 책임이 몰리며 클래스가 비대해지고, 변환 흐름이 단일화되는 부작용도 분명히 존재했습니다.
컨버터를 세분화해 나누는 방식도 고려했지만, 현재 프로젝트의 복잡도와 규모를 고려하면 각 데이터 구조 안에서 변환 책임을 갖도록 구성하는 것이 더 직관적이고 유지보수에 유리하다고 판단했습니다.

따라서 해당 컨버터는 제거하고, 각 클래스에서 변환을 직접 담당하도록 구조를 정리하려 합니다.

감사합니다!

Gomding commented on src/main/java/roomescape/common/domain/DomainId.java:
```diff
@@ -0,0 +1,32 @@
+                .notNullField(Fields.value, value);
+    }
+
+    public Long getValue() {
+        if (assigned) {
+            return value;
+        }
+        throw new IllegalStateException("저장되지 않아서 식별할 수 없습니다.");
+    }
```
값을 요청했을 뿐인데 저장되지 않아서 식별할 수 없다는 상황이 이해할 수 없군요 🤔 
어떤 역할을 하는 메서드일까요?

m-a-king replied:
단순히 null을 방지하기 위한 용도를 넘어서, 식별되지 않은 객체가 비교나 식별 연산에 사용되지 못하도록 하기 위한 제약으로 설계했습니다.

현재 식별자가 없는 상태(unassigned)를 단순히 플래그로만 표시하지 않고, UUID의 상위 비트를 활용해 내부적으로 임시 ID 값을 생성함으로써 null 자체는 방지하고 있습니다. 덕분에 기술적으로는 equals와 hashCode가 정상 작동하며, Set이나 Map에 넣더라도 충돌 없이 사용 가능합니다.

그러나 이 부분에서 모순이 발생한다고 판단했습니다. 임시로 생성된 식별자가 있음으로써 식별되지 않은 객체가 마치 식별 가능한 객체처럼 동작하는 것은 부정확한 표현이며, 오해를 유발할 수 있는 여지가 있다고 생각했습니다.

따라서 이를 명확히 차단하기 위해 getValue() 호출 자체를 차단했습니다. 식별자가 명시적으로 할당되지 않았다면, 어떠한 방식으로든 값을 꺼내어 식별하는 행위 자체를 금지했습니다.
이를 통해 단순 비교뿐 아니라, 식별자의 외부 노출, 컬렉션 내 비교까지 모든 식별 시도를 명시적으로 방지할 수 있게 되었습니다.

결론적으로 이 메서드는 식별자가 할당되지 않은 객체(엔티티)는 어떤 식별 연산에도 참여할 수 없다는 철학을 코드로 명시한 이중 방어적 설계입니다!

Gomding replied:
여전히 `getValue` 만 호출했는데 예외가 발생한다는것은 쉽게 예측하기 어려울 것 같아요 🤔 
팀 컨벤션으로 잘 설명 돼야하는 부분으로 보이네요!

m-a-king replied:
맞습니다. getValue라는 이름이 단순히 값을 가져오는 역할로 보이기 때문에, 예외가 발생하는 흐름은 예측하기 어려울 수 있다는 점에 동의합니다.

getAssignedValue처럼 의도가 좀 더 드러나는 네이밍이었으면 어땠을까 하는 고민도 들었고요.
다만 저는 이 메서드를 설계할 때, 외부에서 객체의 상태를 꺼내서 분기하거나 예외를 던지기보다는, 객체 내부에서 스스로 판단하고 책임을 가지는 구조가 더 좋다고 생각했던 것 같습니다.

물론 그 구조가 단순한 getter 형태에도 동일하게 적용될 수 있는지는 여전히 고민 지점이긴 합니다.

말씀해주신 대로, 이런 판단 기준은 팀의 컨벤션에 따라 조정되는 것이 당연하다고 생각하고, 팀 내에서 혼란을 줄이는 쪽으로 얼마든지 수정할 수 있습니다! 감사합니다.

Gomding commented on src/main/java/roomescape/common/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,34 @@
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler(value = DateTimeParseException.class)
+    public ResponseEntity<String> handleDateTimeParseException() {
```
에러 응답 객체를 만들어보는것은 어떻게 생각하시나요?

m-a-king replied:
RFC 7807에 기반한 표준 오류 응답 형식인 ProblemDetail을 사용하는 방향을 염두에 두고 있었습니다.
다만, 이번 미션은 페어 프로그래밍으로 진행되었고, 그 과정에서 ProblemDetail의 학습과 도입은 미션의 핵심 목표와는 다소 거리가 있다고 판단하여 보류했습니다.

페어 종료 후, 공통 피드백에서도 ProblemDetail에 대한 언급이 있었고, 이에 따라 도입해 볼 계획입니다.
이전에 사용해 본 경험은 있으나, 당시에는 필드를 많이 비워두고 단순하게 사용했던 기억이 있어서 이번에는 보다 의도적으로 구성해 보려 합니다.

만약 ProblemDetail이 불필요하다고 느껴지면, 커스텀 에러 응답 객체를 직접 만들어보겠습니다!

Gomding replied:
> RFC 7807에 기반한 표준 오류 응답 형식인 ProblemDetail을 사용하는 방향을 염두에 두고 있었습니다.

👍 👍 👍 

> 당시에는 필드를 많이 비워두고 단순하게 사용했던 기억이 있어서 이번에는 보다 의도적으로 구성해 보려 합니다.

크으... 💯 💯 💯 

클라이언트와(지금은 강산과 페어가 곧 클라이언트지만) 잘 협의하여 팀에 필요한 형태로 커스텀 에러 응답 객체를 사용하는것 좋습니다!!
저도 표준이라고 하지만 `ProblemDetail` 형태가 유효한가는 의문이 있습니다 😅

m-a-king replied:
감사합니다!
ProblemDetail을 다시 사용해 보면서 이제는 각 필드가 왜 필요한지, 그리고 보편적인 클라이언트 입장에서 어떤 정보들을 기대하는지 조금 더 명확하게 이해할 수 있는 기회가 되었습니다.
물론 제안해 주신 대로, 상황에 따라 클라이언트가 요구하는 스펙에 맞춰 커스텀 에러 응답 객체를 사용하는 것도 충분히 좋은 선택이라고 생각합니다.

Gomding commented on build.gradle:
```diff
@@ -21,11 +21,17 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-jdbc'
+    implementation 'org.springframework.boot:spring-boot-starter-validation'
+
+    implementation 'org.projectlombok:lombok'
+    annotationProcessor 'org.projectlombok:lombok'
```
lombok 을 사용하신 이유가 궁금해요!
lombok 없이 구현할 수는 없었을까요?

m-a-king replied:
Lombok없이 충분히 구현할 수 있지만 getter, 생성자, equals, hashCode, toString 등 반복적인 보일러플레이트 코드를 줄여 생산성과 가독성을 높이기 위해 사용했습니다.

또, 전체적으로 Lombok을 활용하다가 Lombok 없이 직접 메서드를 구현한 경우에는 "이 메서드는 왜 직접 구현되었을까?"라는 질문을 유도하고, 그 안에 담긴 특정 로직이나 제한 의도를 코드로 명확히 드러낼 수도 있었습니다. (예: 단순 조회를 넘는 로직, 값 접근 제한 등)

즉, Lombok은 일반적인 경우를 간결하게 처리하고, 예외적인 상황은 명시적으로 표현할 수 있는 대비 효과가 있어 의미 있다고 판단했습니다.

Gomding replied:
Lombok 을 사용하면서 발생할 수 있는 예측못하는 에러나 버전 호환등의 이슈는 어떻게 생각하시나요?
저는 실무에서 Lombok 때문에 kotlin 전환이 곤란헀던 경험도 있었어요 :)

m-a-king replied:
말씀해 주시니, 저도 예전에 간단한 토이 프로젝트를 코틀린으로 전환해 보면서, 여러 어노테이션 관련 문제가 발생했던 기억이 납니다.

마침 답글을 작성하는 지금, 코치님이 옆에서 여러 키워드를 언급해 주셨고 관련 개념을 더 깊게 찾아볼 수 있었고, 이 과정에서 아래 링크의 좋은 정보를 얻을 수 있었습니다,
[좋은 정보들](https://d2.naver.com/helloworld/6685007)

Lombok은 컴파일 타임에 어노테이션 프로세서를 통해 바이트코드를 수정하는 방식이라, IDE에서 제대로 인식하지 못하거나 디버깅이 어려운 경우가 있었고, 버전에 따라 예상치 못한 호환성 문제가 발생할 수 있다는 점도 새롭게 알게 되었습니다.

Lombok, 이거 써도 되는 걸까? 하는 생각이 들기도 했지만, 실제로 사용해 보면서 가독성이 크게 개선되고, 보일러 플레이트를 줄일 수 있다는 점은 여전히 큰 장점이라고 느낍니다.
특히 저는 복잡한 기능보다는 반복적인 코드를 줄이는 데 집중하고 있어서, 아직까지는 Lombok을 사용하는 쪽을 선택하고 싶습니다.

또, 이제는 그 단점들도 명확히 인지하고 있고, 그렇기 때문에 안정적인 LTS 조합에서 더 자신 있게 Lombok을 사용하자고 설명할 수 있을 것 같습니다!

감사합니다.

Gomding commented on src/main/java/roomescape/reservation/domain/ReservationId.java:
```diff
@@ -0,0 +1,20 @@
+        super(value, assigned);
+    }
+
+    public static ReservationId unassigned() {
+        return new ReservationId(UUID.randomUUID().getMostSignificantBits(), false);
+    }
```
영속화 전에 임시 ID 를 사용하시는군요
어떤 장점이 있어서 사용하시는 걸까요?

m-a-king replied:
임시 ID는 실제로 사용될 수는 없지만, 식별되지 않은 객체들끼리 동등성 비교 시 오인되는 문제를 방지하기 위해 null 대신 UUID 기반의 임시 값을 부여했었습니다. 또, null은 값이 없다는 특성상 동등성 비교뿐 만 아니라, 출력, 컬렉션 사용 등에서 부작용을 유발할 수 있다고 판단했고, 이를 명시적으로 차단하고자 했습니다!

[이전 설명](https://github.com/woowacourse/spring-roomescape-member/pull/222#discussion_r2072389955) 에서 자세하게 다뤘습니다.

Gomding replied:
equals & hashCode 재정의 시 id가 null일때는 동등하지 않다고 판단해주면 되지 않을까요? 🤔 
단순 동등성 비교때문에 구현했다는 것은 구현 공수가 너무 큰 것 같아요!

m-a-king replied:
코드를 다시 살펴보니, DomainId를 상속받는 하위 타입마다 정적 팩토리 메서드에서 UUID를 반드시 생성해 전달해야 했고, 이로 인해 새로운 ID 클래스를 추가할 때마다 중복된 생성 책임이 반복되는 구조였습니다. 이를 null로 두는 방식으로 전환하면 하위 타입에서 생성 방식에 대한 제약 없이 조금 더 유연하게 설계할 수 있을 것이라 판단했습니다.

피드백 해주신대로, DomainId의 equals와 hashCode 구현을 value가 null일 경우에는 항상 동등하지 않다고 간주하는 방식으로 수정하겠습니다. 감사합니다!

Gomding commented on src/main/java/roomescape/theme/domain/Theme.java:
```diff
@@ -0,0 +1,52 @@
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+@FieldNameConstants(level = AccessLevel.PRIVATE)
+@EqualsAndHashCode(of = "id")
+public class Theme {
```
Theme 객체에 대한 테스트는 작성하지 않아도 괜찮을까요?

m-a-king replied:
곰곰이 생각해봤을 때, 현재 Theme 객체에는 별도의 도메인 로직이 없고, 유효성 검사는 Validator 클래스를 통해 별도로 테스트되고 있어 테스트의 필요성을 크게 느끼지 못했습니다.

하지만 어디를 테스트해야 할지 곰곰이 살펴보던 중, withId()라는 명확한 의도를 가진 정적 팩토리 메서드를 사용하는 경우에는 내부적으로 unassigned ID가 들어가는 상황을 방지해야 할 필요성이 있다고 느꼈습니다. withId라면 실제 식별자가 존재해야 한다는 암묵적인 계약이 있으므로, 이 부분은 명시적으로 제약을 걸어주는 것이 맞겠다고 판단했습니다. 물론 이는 Theme 도메인에만 해당하는 것이 아니라 전체 도메인 객체 생성 규칙 전반에 적용하겠습니다!

m-a-king replied:
[코드 래빗](https://github.com/woowacourse/spring-roomescape-member/pull/222#discussion_r2072703912)

이 테스트가 필요했습니다... 😢 
 반영하겠습니다!

Gomding replied:
Validator 를 통해 별도로 테스트 된다는것을 이해하지 못했어요!
다른 사람이 Validator 코드 또는 Theme 클래스 내부의 `validate` 메서드를 일부 훼손하거나 잘못 수정하면 
Theme 객체가 정상 규칙에 의해 정상 생성되는걸 보장할 수 있나요?
테스트코드가 없다면 빠르게 피드백 받을 수도 없을 것 같아요!

m-a-king replied:
피드백 주신 내용에 동의합니다!
처음엔 동일한 검증 로직을 반복하는 것처럼 보여 중복이라고 생각했지만, 곰곰이 생각해 보니 전혀 다른 성격의 테스트라는 걸 알게 되었습니다.

실제로 이 부분에 대한 테스트가 없다면, 문제가 발생했을 때 빠르게 감지하기 어려울 수 있다는 점도 중요한 포인트라고 느꼈습니다.
충분히 테스트할 가치가 있는 부분이라 생각되어, 말씀 주신 방향대로 추가하도록 하겠습니다. 감사합니다!

Gomding commented on src/test/java/roomescape/reservation/domain/ReserverNameTest.java:
```diff
@@ -0,0 +1,26 @@
+    @Test
+    @DisplayName("예약자 이름은 비어있을 수 없다")
+    void cannotEmptyReserverName() {
+        // given
+        // when
+        // then
+        assertAll(() -> {
```
별도의 given 이 없다면 다음과 같이 표현하셔도 좋아요
가독성을 위한 것이니까요

```suggestion
        // when then
        assertAll(() -> {
```

m-a-king replied:
개인적으로는 given, when, then을 모두 명시하는 방식이 이후 given에 내용이 추가될 때 더 유연하게 대응할 수 있었고, 특히 given에 아무 내용이 없다는 점도 명확히 드러나 스스로 코드를 점검해 보는 계기가 되었습니다.
라이브 템플릿을 활용해 작성하다 보니 코드를 입력하는 데에는 큰 불편이 없었지만, 파일 크기가 커짐에 따라 유지 보수성이 떨어지고, 가독성이 떨어져서 리뷰하실 때 불편함을 드릴 수 있겠다는 생각도 들었습니다.
앞으로는 권장해 주시는 방식대로 표현해겠습니다!

Gomding commented on src/test/java/roomescape/reservation/application/usecase/ReservationQueryUseCaseImplTest.java:
```diff
@@ -0,0 +1,134 @@
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+@SpringBootTest
+@Transactional
+class ReservationQueryUseCaseImplTest {
```
테스트에 `@Transactional` 을 사용해주셨군요
해당 애너테이션 사용 시 장단점은 무엇일까요?

m-a-king replied:
테스트 클래스에서 `@Transactional`을 사용하면 각 테스트가 독립적인 트랜잭션 내에서 실행되고, 테스트 종료 시 자동으로 롤백되기 때문에 데이터 정리 부담 없이 테스트 간 영향을 최소화할 수 있다는 장점이 있었습니다.

이후에 JPA에서는 트랜잭션 안에서만 엔티티의 변경 사항이 DB에 반영되는 flush가 안전하게 작동하고, save 이후 findById 와 같은 흐름도 실제 DB 반영을 검증하는 데 있어 신뢰성이 높아지기 때문에, 변경 감지나 영속성 전이 등의 테스트를 안정적으로 수행할 수 있다고 학습했습니다.

다만 테스트에만 트랜잭션이 선언되어 있고 실제 프로덕션 코드에서는 누락된 경우, 테스트는 통과하더라도 운영 환경에서는 flush나 commit이 발생하지 않아 DB에 변경 사항이 반영되지 않는 문제가 발생할 수 있으므로 주의가 필요하다고 생각했습니다.

이전에 토비님, 향로님, 제미니님의 사례를 통해 인상 깊게 접해서 기억에 남아 있습니다!

Gomding commented on src/main/java/roomescape/common/validate/Validator.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    public Validator notNullField(final String fieldName, final Object target) {
+        if (target == null) {
+            throw new IllegalArgumentException(clazz.getSimpleName() + "." + fieldName + " 은(는) null일 수 없습니다.");
+        }
```
예외에 클래스 이름을 포함하는 이유가 궁금해요 🤔 
서버에서 필요하다면 예외 스택에 있을 것이고
클라이언트에서 필요할까 생각해보면 관심없는 정보일 것 같아서요

m-a-king replied:
맞습니다😭  

이 부분은 제출 이후 서둘러 리팩토링 한 영역 중 하나였습니다.
수정을 진행하면서도 예외 처리는 여전히 완벽하게 해내기 어려운 영역이라는 걸 많이 느꼈습니다.

처음부터 예외 메시지를 상세히 설계하고 진행하는 것이 좋을지, 아니면 실제 클라이언트와 소통하면서 "특정 예외 메시지가 이상하다"는 피드백을 받아가며 점진적으로 개선해나가는 것이 나을지 고민이 됩니다.

또 프론트엔드 측에서도 저희가 내려주는 메시지를 그대로 사용자에게 보여주기보다는 대부분 자체적으로 가공해서 UI에 띄우는 경우가 많다 보니, 클라이언트에게 전달되는 메시지도 차라리 영어로 간단하게 처리하는 게 좋지 않을까? 하는 생각도 들었습니다.

조금 더 찾아보니, 이 부분은 팀의 합의와 도메인 특성에 따라 방식이 많이 달라져서, 저 역시 앞으로도 계속 고민하고 개선해나갈 주제라고 생각합니다!

---
 
> 서버에서 필요하다면 예외 스택에 있을 것이고

이 또한 맞습니다. 예외 스택에 클래스 이름이 포함되지만, 예외 메시지에 클래스와 필드명을 명시한 이유는 로깅 시점에서 메시지만 보고도 바로 문제를 파악할 수 있도록 하기 위함이었습니다.

스택 트레이스를 통해서도 확인할 수 있지만, 로그만 보고도 빠르게 진단이 가능하면 디버깅 속도가 높아진다고 생각했습니다. 특히 서버 로그를 탐색하거나, 모니터링 도구로 에러 알림을 받을 때 메시지만 노출되는 경우, 클래스와 필드 정보를 함께 포함하면 더 유용하다고 느꼈습니다. 

찰리의 피드백을 곰곰히 생각해 보니, 중복된 정보일까 고민하게 되는 것 같습니다. 🤔

Gomding replied:
더욱 친절한 메세지는 언제나 좋지만
중복된 정보는 꼭 필요할까 라는 생각입니다 :)

로그만 보고 추적가능한 경우가 있지만
보통 예외 스택을 보고 호출처를 따라가보는 경우가 더 많다고 생각해요 :)
특히 현재 상황은 어디서 null 이 들어온건지 찾아다닐거구요 ㅎㅎ

강산과 페어 혹은 팀이 디버깅하는데 더 유용하다고 판단한다면 OK 라고 생각합니다! 👍

m-a-king replied:
> 어디서 null 이 들어온건지 찾아다닐거구요 ㅎㅎ

맞습니다. 결국 null이 어디서 들어왔는지를 찾아다닐 것 같습니다.

---

감사합니다. 예외 처리에 대해서는 앞으로도 지속적으로 고민하면서 트레이드오프를 신중히 따져보고, 저만의 판단 기준을 세워나가겠습니다. 😄

Gomding reviewed (changes requested):
안녕하세요 강산!
방탈출 사용자 예약 미션 함께하게된 찰리입니다
몇가지 코멘트 남겼으니 확인부탁드려요~

궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요!!

m-a-king committed:
refactor: Converter 클래스 삭제, 각 클래스에서 변환 담당

m-a-king committed:
refactor: DomainId getter 예외 메시지 수정

m-a-king committed:
refactor: 원시값 포장 (예약 수, 예약 여부)

m-a-king committed:
refactor: 예외 처리 개선

- 도메인 이름 상수 추가 및 검증 오류 메시지 개선
- 커스텀 예외 처리 추가 및 기존 예외를 커스텀 예외로 변경
- 로깅 추가

m-a-king committed:
refactor: InternalLogicException 제거 및 IllegalArgumentException으로 변경

m-a-king committed:
feat: 식별자 할당 확인 로직 추가

- Theme, Reservation, ReservationTime 등에서 requireAssigned 호출 추가

m-a-king committed:
refactor: 변수 명 변경

m-a-king committed:
feat: 유즈케이스 클래스에 트랜잭션 어노테이션 추가

m-a-king committed:
refactor: 테마 쎔네일 필드 URI로 변경, Validator 클래스 개선

m-a-king committed:
refactor: 공통 예외 처리 및 도메인 용어 관리 개선

- 기존 공통적인 예외를 통합
- 도메인에서 부분적으로 사용되던 domainName을 DomainTerm ENUM으로 대체하여 코드 간결화
- Theme 관련 쿼리에서 존재 여부를 확인하는 existsById, existsByName 메서드 추가

m-a-king committed:
test: 존재하는 테마 이름 예외 검증 테스트

m-a-king committed:
refactor: 변수명 개선

m-a-king committed:
feat: 예외 클래스에 HttpStatus code 추가

m-a-king committed:
feat: 제약 충돌 예외 클래스 추가

m-a-king committed:
refactor: 예외 처리 일관성 개선

m-a-king committed:
refactor: 시간 의존 로직을 TimeProvider로 위임하여 LocalTime.now 직접 호출 제거

m-a-king commented:
# 추가 질문이 없습니다!

# [리뷰 이후 변경 사항](https://github.com/woowacourse/spring-roomescape-member/pull/222/files/a4abd3d287f7bd6d53abd72bbbec7146c57da17f..af3caccdfb3b3139724c57a3f47760033cab79c0)

# [리뷰 이전 변경 사항 + 리뷰 이후 변경 사항](https://github.com/woowacourse/spring-roomescape-member/pull/222/files/77d35d6d545a9f5a9ecd27c7d473bf342936f5ff..af3caccdfb3b3139724c57a3f47760033cab79c0)

Gomding commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -0,0 +1,83 @@
+            return;
+        }
+
+        if (date.isBefore(now.toLocalDate())) {
+            throw new PastDateReservationException(date, now);
+        }
+
+        if (time.isBefore(now.toLocalTime())) {
+            throw new PastTimeReservationException(time, now);
+        }
```
👍 👍 👍

m-a-king replied:
😄

Gomding commented on src/main/java/roomescape/common/time/TimeProvider.java:
```diff
@@ -0,0 +1,28 @@
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+
+@Component
+@AllArgsConstructor
+public class TimeProvider {
+
+    private final Clock clock;
```
시간 생성 책임을 가진 객체를 빈으로 주입하도록 만들어주셨으니 
시간 관련 테스트도 쉽게할 수 있겠네요 👍 👍

m-a-king replied:
로직에서 LocalDateTime을 직접 사용하는 것이 어딘가 모르게 불편하게 느껴져서 추출했습니다!
예전에 GitHub Actions나 클라우드 환경에서 아무 설정 없이 LocalDateTime을 그대로 썼다가 문제가 터진 기억 때문인지, 이번에도 알 수 없는 두려움이 살짝 올라오더라고요. 😨

Gomding commented on src/test/java/roomescape/reservation/application/usecase/ReservationCommandUseCaseImplTest.java:
```diff
@@ -0,0 +1,214 @@
+                );
+    }
+
+    @Test
+    @DisplayName("지나간 날짜/시간에 대한 예약을 생성할 수 없다")
+    void cannotCreatePastDateTimeReservation() {
```
테스트가 실패하고 있네요!

m-a-king replied:
제 환경에서는 정상적으로 실행되었지만, 어떤 문제가 있을지 살펴보니 23시 n분에 +1시간 또는 00시 n분에 -1시간을 적용하는 테스트 코드에서 오류가 발생했을 것으로 예상합니다.
테스트 실패를 방지하기 위해, 1나노초 단위로 값을 조정하여 최대한 안정적인 테스트가 되도록 수정하겠습니다.
물론 Clock 등을 활용해 시간을 고정하는 방법도 고려할 수 있지만, 현재 상황에서는 굳이 필요하지 않다고 판단했습니다!

Gomding reviewed (approved):
안녕하세요 강산!
깊은 고민이 느껴지네요 👍 👍 
이번 단계는 충분히 구현하신 것 같아서 머지할게요 😃 
추가로 남긴 코멘트는 다음 단계에서 함꼐 고민해주세요~!

추가로 궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요~
